qwen-3-8b-agent (#3103)
huggingface/blog: intel-qwen3-agent.md

---
title: "Accelerating Qwen3-8B Agent on Intel® Core™ Ultra with Depth-Pruned Draft Models" 
thumbnail: /blog/assets/intel-qwen3-agent/smolagents2.png
authors:
- user: imargulis
  guest: true
  org: Intel
- user: ofirzaf
  guest: true
  org: Intel  
- user: sguskin
  guest: true
  org: Intel  
- user: guybd
  guest: true
  org: Intel
- user: pcuenq
---

> ****
> # ![banner image](assets/intel-qwen3-agent/smolagents2.png)

Reviewer: ```suggestion
![banner image](/blog/assets/intel-qwen3-agent/smolagents2.png)
```

# Accelerating Qwen3-8B Agent on Intel® Core™ Ultra with Depth-Pruned Draft Models

TL;DR:

-   [Qwen3-8B](https://qwenlm.github.io/blog/qwen3/) is one of the most exciting recent releases—a model with  native agentic capabilities, making it a natural fit for the AIPC.

-   With [OpenVINO.GenAI](https://github.com/openvinotoolkit/openvino.genai), we’ve been able to accelerate generation by ~1.3× using speculative decoding with a lightweight Qwen3-0.6B draft.

> -   By applying a simple pruning process to the draft, we pushed the speedup even further to ~1.4×

Reviewer: ```suggestion
-   By using speculative decoding and applying a simple pruning process to the draft, we pushed the speedup even further to ~1.4×
```

Suggesting to add a bit of context regarding what "draft" refers to.

-   We wrapped this up by showing how these improvements can be used to run a fast, local AI Agent with 🤗[smolagents](https://github.com/huggingface/smolagents)

## Qwen3
Qwen3-8B is part of the latest Qwen family, trained with explicit agentic behaviors. It supports tool invocation, multi-step reasoning, and long-context handling capabilities, that make it well-suited for complex agent workflows. When integrated with frameworks like Hugging Face 🤗smolagents, QwenAgent, or AutoGen, it enables a wide range of agentic applications built around tool use and reasoning.
The combination of optimized inference and built-in agentic intelligence makes Qwen3-8B a compelling foundation for next-gen AI agents.

## Accelerating Qwen3-8B on Intel® Core™ Ultra with Speculative Decoding

> We started by benchmarking the 4-bit optimized OpenVINO version of Qwen3-8B on an Intel Lunar Lake integrated GPU, establishing this as our baseline for further acceleration

Reviewer 2: Added

[Speculative decoding](https://arxiv.org/abs/2211.17192) is a method to speed up auto-regressive generation. It works by using a smaller, faster model as a draft to propose multiple tokens in a single forward pass, which are then validated by the larger target model in one forward pass. In our setup, [Qwen3-8B](https://huggingface.co/Qwen/Qwen3-8B) served as the target model while [Qwen3-0.6B](https://huggingface.co/Qwen/Qwen3-0.6B) was used as the draft. This approach delivered an average of 1.3× speedup over the baseline.

```python
from openvino_genai import LLMPipeline, draft_model

> target_path = "/path/to/target/model"
> 
> draft_path = "/path/to/draft/model"
> 
> device = "GPU"
> 
> model = LLMPipeline(target_path, device, draft_model=draft_model(draft_path, device))
> 
> streamer = lambda x: print(x, end="", flush=True)
> 
> model.generate("What is speculative decoding and how does it improve inference speed?", max_new_tokens=100, reamer=streamer)

Reviewer: ```suggestion
target_path = "/path/to/target/model"
draft_path = "/path/to/draft/model"
device = "GPU"

model = LLMPipeline(target_path, device, draft_model=draft_model(draft_path, device))

streamer = lambda x: print(x, end="", flush=True)
model.generate("What is speculative decoding and how does it improve inference speed?", max_new_tokens=100, reamer=streamer)
```

```

## Pushing Performance Further

>  The speedup from SD depends on the mean number of generated tokens per forward step of the target, $\gamma$ the speculation window size, and the ratio between the target and draft models' latency $c$. A smaller, faster (though less accurate) draft can often deliver greater acceleration. This inspired us to shrink the draft model while still preserving its quality, i.e. $E(\\# generated\ tokens)$.

Reviewer: ```suggestion
 The speculative decoding speedup depends on the average number of generated tokens per forward step of the target, \\( \gamma \\), the speculation window size, and the ratio between the target and draft models' latency \\( c \\). A smaller, faster (though less accurate) draft can often deliver greater acceleration. This inspired us to shrink the draft model while still preserving its quality, i.e. \\( E(\# generated\_tokens) \\).
```

$$

> Speedup = \frac{E(\\# generated\ tokens)}{\gamma c + 1}

Reviewer: ```suggestion
Speedup = \frac{E(\# generated\_tokens)}{\gamma c + 1}
```

$$

> Our [recent work](https://huggingface.co/papers/2411.11055) shows that model depth (number of layers) is a major contributor to inference latency. Building on insights from [Gromov et al.](https://huggingface.co/papers/2403.17887) and [Siddiqui et al.](https://arxiv.org/abs/2407.16286), who showed that layer pruning can be done with minimal accuracy loss, we pruned 6 of 28 layers from the Qwen3-0.6B draft.

Reviewer: Consider using less inline citations to make the text more approachable to a wider audience. We can add all references as resources at the end of the post.

Reviewer 2: Fixed

> Our [recent work](https://huggingface.co/papers/2411.11055) shows that model depth (number of layers) is a major contributor to inference latency. Building on insights from [Gromov et al.](https://huggingface.co/papers/2403.17887) and [Siddiqui et al.](https://arxiv.org/abs/2407.16286), who showed that layer pruning can be done with minimal accuracy loss, we pruned 6 of 28 layers from the Qwen3-0.6B draft.

Reviewer 3: thanks!

To recover the quality of the pruned draft model, we further finetuned it using synthetic data generated by Qwen3-8B.
The data was produced by generating responses to 500k prompts from [BAAI/Infinity-Instruct dataset](https://huggingface.co/datasets/BAAI/Infinity-Instruct).

The resulting pruned draft model delivered \~1.4x speedup compared to the baseline, an improvement over the \~1.3× gain achieved with the original draft. This outcome aligns with theoretical expectations - reducing draft latency improves the over-all speedup, enabling faster and more efficient inference.

This demonstrates how pruning + speculative decoding can unlock faster and more efficient inference—making local AI agents even more practical.

Check out the [notebook](https://github.com/guybd/openvino_notebooks/blob/latest/supplementary_materials/notebooks/qwen-3/qwen3.ipynb) and the Qwen3-0.6B depth-pruned [draft model](https://huggingface.co/OpenVINO/Qwen3-pruned-6L-from-0.6B-int8-ov) to reproduce our results step by step

> ## Integration to 🤗smolagents

Reviewer: ```suggestion
## Integration with 🤗smolagents
```

The benefits aren’t limited to Hugging Face, this model pairing can also be used seamlessly with frameworks like AutoGen or QwenAgent, further strengthening the agentic ecosystem.

In our demo, we assigned the accelerated Qwen3-based agent a task: 👉 Summarize the key features of the Qwen3 model series and present them in a slide deck.

Here’s how it worked:
1.	The agent used a web search tool to gather up-to-date information.
2.	It then switched to the Python interpreter to generate slides with the `python-pptx` library.
This simple workflow highlights just a fraction of the possibilities unlocked when accelerated Qwen3 models meet frameworks like 🤗smolagents, bringing practical, efficient AI agents to life on AI PC. Try it [here](https://github.com/guybd/openvino_notebooks/blob/latest/supplementary_materials/notebooks/qwen-3/smolagents/qwen3_agent.ipynb) 🚀

> <iframe width="100%" style="aspect-ratio: 16 / 9;"src="https://youtu.be/irSDd5lNxIk" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

Reviewer: ```suggestion
<iframe width="100%" style="aspect-ratio: 16 / 9;"src="https://www.youtube.com/embed/irSDd5lNxIk" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```

I think we need an embed URL.

---
title: "Accelerating Qwen3-8B Agent on Intel® Core™ Ultra with Depth-Pruned Draft Models" 

> thumbnail: /blog/assets/intel-qwen3-agent/smolagents2.png

Reviewer 4: ```suggestion
thumbnail: /blog/assets/intel-qwen3-agent/smolagents-1300-650.png
```

authors:
- user: imargulis
  guest: true
  org: Intel
- user: ofirzaf
  guest: true
  org: Intel  
- user: sguskin
  guest: true
  org: Intel  
- user: guybd
  guest: true
  org: Intel
- user: pcuenq
---

> ![banner image](/blog/assets/intel-qwen3-agent/smolagents2.png)

Reviewer 4: ```suggestion
![banner image](/blog/assets/intel-qwen3-agent/smolagents-1300-650.png)
```
